### DISCUSSION TITLE
[plugin] request collapsing (service aggregation)

### DISCUSSION THREAD

**Discussion by sonicaghi** [Kong/kong#7350] [Ideas and feature requests]
Example: https://github.com/Netflix/Hystrix/wiki/How-To-Use#Collapsing

**Comment by ahmadnassri** [comment]
more: http://microservices.io/patterns/apigateway.html

this is traditionally referred to as "batching proxy". basically client sends a multi-part request, representing all the requests it wants to get, and the proxy completes those requests and responds with a multi-part response.

in the case of kong the multi-part request can be configured in the plugin / kong side, and the response is a multi-part response.

**Comment by lsmith77** [comment]
+1

**Comment by lindo-jmm** [comment]
@ahmadnassri Would it be possible to transform the response before passing it back to the client and also targeting devices? It's understood that everything is still in the idea phase.

**Comment by ahmadnassri** [comment]
@lindo-jmm depends on what you mean by "targeting devices" ?

**Comment by heat** [comment]
I would love Kong do that but I wondering if its responsibility.  https://encrypted-tbn3.gstatic.com/images?q=tbn:ANd9GcQ35pXZMfL-YNeA67_FbT8uxvhgz1_1ONpXiTddYzjn9mnbycd3bU-IS11ZbQ there need a edge service in our api to compose load balance e fault tolerance enhance

**Comment by brunosimioni** [comment]
+1

**Comment by avidyalalala** [comment]
+1

**Comment by tshwangq** [comment]
+1

**Comment by blackxored** [comment]
+1

**Comment by subnetmarco** [comment]
In order to start working on this plugin, could you provide some examples of requests you would like to collapse, and how the response of those requests should be aggregated in one output?

**Comment by ahmadnassri** [comment]
@thefosk 

there are two design patterns here that I know of:

### simple:

plugin definition defines multiple upstream paths to query, and respond back with a single multi-part response that includes all results (headers + bodies)

an upstream timeout configuration is critical here, _(preferably a timeout per upstream)_.

### complex

> this is traditionally referred to as "batching proxy". basically client sends a multi-part request, representing all the requests it wants to get, and the proxy completes those requests and responds with a multi-part response.

**tl;dr:**

> multi-part in => multi-part out.

**detailed example**
incoming request, contains a body (multi-part), each part includes headers related to the target upstream (headers, path, etc ...)

``` http
GET /my/collapsed/api HTTP/1.1
Host: localhost:8000
Accept: multipart/form-data
Connection: keep-alive
Content-Type: multipart/form-data; boundary=---------------------------xxx
Content-Length: 554

-----------------------------xxx
Host: api.foo.com
Content-Type: application/json
Content-Length: 14

{"foo": "bar"}
-----------------------------xxx
Host: api.foo.com
Content-Type: application/xml

<xml><foo>bar</foo></xml>
-----------------------------xxx
Host: api.foo.com
Content-Type: text/html

<!DOCTYPE html><title>Content of a.html.</title>

-----------------------------xxx--
```
- can use `Host` in each part, but better to be a custom Kong header, depicting upstream API name, identifier, etc ...

outgoing response is a multipart response detailing each of the upstream results

**Comment by Tieske** [comment]
I think this requires some more details before implementing. Eg. the example 
given above lacks context as to which response came from which upstream.

As for implementation;
within nginx we have one request process, but we need multiple responses. If we 
do all the upstream requests from the plugin, we'll be bypassing several process 
steps in the nginx process (eg. bypass logging phase).
So probably we should have one request being handled by nginx in the request
process ('main request'), and the others via additional requests from Lua 
code ('sub requests').

This will have the side effect that the sub requests can only exploit
the dns based loadbalancing, whereas the main request will be able to
use the 'upstreams' balancer, and use all other api defaults for example.

To make it play nicely with other plugins, this plugin would probably have to
run last (plugin 'priority') so all other plugins have been executed
already (plugins running after this one will only operate on the main request, 
and not on the sub requests)

How to deal with errors?
 - local error (eg. timeout)
 - upstream error (eg. 404)

How to pass parameters in the request? does every upstream get all parameters?
for now: identical query-string and body goes to every upstream

What http methods should it support?
for now: GET only, anything else gets a 405


__Configuration__:
```lua
{
  main_name = "some name",     -- equals the names below, but for the main request
  sub_requests = {             -- array of upstream request targets
    {
      name = "upstream2",
      timeout = 10,
      url = "http://some/where/",
    },{
      name = "upstream3",
      timeout = 10,
      url = "http://some/where/",
    },
  },
  response = "application/json"    -- response type of the combined response (json or url)
  header_prefix = "X-Kong-batch-", -- for url encoded responses
}
```

__JSON response__:
status: always 200
body;
```
{
  "some name": {                     -- good-path, proper upstream result
      "status": xxx,                 -- the actual status code
      "body": {}                     -- nested json body if it was json, string otherwise
    },
  "upstream1": {
      "status": 500                  -- internal Kong error (eg. timeout)
    },
  "upstream2": {
      "status": 404,                 -- upstream error; basically same as good-path
      "body": {}                     -- nested json body if it was json, string otherwise
    }
}
```

__Multipart response__:
status: always 200
body;
as mentioned above, but each part will receive extra headers based on 'header_prefix';
eg. `X-Kong-batch-name=some%20name` and `X-Kong-batch-status=200`.
An internal error will result in `X-Kong-batch-status=500` with an empty body.
The order of the parts should be identical to the order of the main/sub url's


Does the above make sense?

**Comment by liu0013** [comment]
+1

**Comment by jondavidford** [comment]
+1

**Comment by liubog2008** [comment]
Is there any progress?

**Comment by bryan-lai** [comment]
+1 waiting for this plugin

**Comment by jrbarros** [comment]
me too

**Comment by sanketgandhi** [comment]
+1 waiting

**Comment by YGwen** [comment]
+1

**Comment by Moxxie** [comment]
Any status on this?

**Comment by smileMrLee** [comment]
+1

**Comment by billyteves** [comment]
Any update? :)

**Comment by subnetmarco** [comment]
Hi everybody - This plugin can get very complex since everybody has very unique requirements. If you could share with us your use-case it would help us determine what's the best approach for this plugin. The more use-cases we can collect the better.

As of today we have been working on this functionality on a case by case basis with our enterprise customers, but maybe there are ways to implement a generic enough interface.

**Comment by orientzc** [comment]
+1 waiting

**Comment by TomFreeman** [comment]
@thefosk 

We have a relatively (ahem) simple use-case - although it sounds quite different to the original one, and maybe it's already resolved elsewhere.

Essentially, we want a single call to fan-out into multiple API calls, and then to combine the returned document.

Something like this:

```
GET /my/aggregate HTTP/1.1
Host: localhost:8000
Accept: application/json
```
Fans out to multiple calls:
```
Host: api.foo.com
Content-Type: application/json
Content-Length: 14

{"foo1": "foo", "bar2": "bar bar"}
-----------------------------xxx
Host: api.bar.com
Content-Type: application/xml

{"foo2": "foo foo", "bar2": "bar bar"}
```

Which gives the resulting document:
```
{"foo1": "foo", "foo2": "foo foo", "bar2": "bar bar"}
```

Preferably we'd like to be able to combine nested object trees, and possibly in some sort of defined order.

Ultimately we'd like to declare some of the fan-out calls as optional, so the entire call can succeed without them.

Would we have to write our own plugins to do that at this point?

**Comment by shaowin16** [comment]
Where can I find the plugin 'request collapsing' ?

**Comment by Tieske** [comment]
@shaowin16 it hasn't been written yet. So far it is a proposal/request for a new plugin.

**Comment by liuyanzhi08** [comment]
+1 waiting

**Comment by jharitesh108** [comment]
+1 waiting

**Comment by phamtienhung** [comment]
+1

**Comment by withlin** [comment]
+1 waiting

**Comment by caiorondon** [comment]
+1 waiting

**Comment by sunilb20** [comment]
+1 Waiting...

Use case for us : 
1. For client, there is single gateway api call having sufficient input data as JSON.
2. Gateway then calls multiple backend api's and should support both sequential and parallel calls.
3. In case of Sequential calls, the response from 1st microservice should be passed to 2nd and so on. There may be a need of transformation to pass only applicable attributes instead of all. Also responses from each  microservices should be preserved to aggregate at the end.
4. In case of parallel calls, there must be a transformation stage for each back end call to pass the relevant fields from the request json. At the end of completion of all the calls, they all should join and should have a way to aggregate based on the client requirements.

**Comment by ankurmalik22** [comment]
+1 for the above use case, only change I want to request is to please add support for additional message formats such as JSON, XML etc

1. For the client, there is a single gateway API call having sufficient input data as JSON/XML.

**Comment by njuptxiaot** [comment]
+1

**Comment by carnei-ro** [comment]
Hi guys, I've made something close to that:
https://bitbucket.org/leandro-carneiro/kong-aggregator/src/master/

**Comment by mittaliavi** [comment]
> @thefosk
> 
> We have a relatively (ahem) simple use-case - although it sounds quite different to the original one, and maybe it's already resolved elsewhere.
> 
> Essentially, we want a single call to fan-out into multiple API calls, and then to combine the returned document.
> 
> Something like this:
> 
> ```
> GET /my/aggregate HTTP/1.1
> Host: localhost:8000
> Accept: application/json
> ```
> Fans out to multiple calls:
> 
> ```
> Host: api.foo.com
> Content-Type: application/json
> Content-Length: 14
> 
> {"foo1": "foo", "bar2": "bar bar"}
> -----------------------------xxx
> Host: api.bar.com
> Content-Type: application/xml
> 
> {"foo2": "foo foo", "bar2": "bar bar"}
> ```
> Which gives the resulting document:
> 
> ```
> {"foo1": "foo", "foo2": "foo foo", "bar2": "bar bar"}
> ```
> Preferably we'd like to be able to combine nested object trees, and possibly in some sort of defined order.
> 
> Ultimately we'd like to declare some of the fan-out calls as optional, so the entire call can succeed without them.
> 
> Would we have to write our own plugins to do that at this point?

Any idea , if we have this plugin available now?

**Comment by Tieske** [comment]
> Hi guys, I've made something close to that:
> https://bitbucket.org/leandro-carneiro/kong-aggregator/src/master/

@carnei-ro can you add a license to your repo?

**Comment by carnei-ro** [comment]
@Tieske I moved from bitbucket to github, please, use the link: https://github.com/carnei-ro/kong-aggregator. It has MIT license.

**Comment by BearKid** [comment]
May these help ?
- [Json Batching of Microsoft Graph](https://docs.microsoft.com/en-us/graph/json-batching) 
- [OData JSON Format Version 4.01 specification -  Batch Requests and Responses](http://docs.oasis-open.org/odata/odata-json-format/v4.01/odata-json-format-v4.01.html#sec_BatchRequestsandResponses)

**Comment by majuansari** [comment]
@subnetmarco  @Tieske Any plans to implement this officially in Kong (Enterprise or OSS)? Or is this something the community could develop with Go plugins?

**Comment by vgrund** [comment]
@subnetmarco @Tieske I would like to have any news about this implementation.

**Comment by jharitesh108** [comment]
waiting for this awesome feature

**Comment by kkmoslehpour** [comment]
We are also waiting for this feature

**Comment by carnei-ro** [comment]
Hey guys, this is a really old thread. I made that plugin a long time ago. I could update it if you tell me some uses cases….

**Reply by kkmoslehpour** [reply]
Could you please point me to the plugin? If there was an easy way to allow Kong to make request to multiple upstream services APIs to aggregate responses similar to [Tyk's compose API feature](https://tyk.io/docs/advanced-configuration/compose-apis/). This is a key feature of what API gateways should provide https://docs.microsoft.com/en-us/azure/architecture/patterns/gateway-aggregation

**Reply by carnei-ro** [reply]
https://github.com/carnei-ro/kong-aggregator

It is a really old and not maintained code

**Comment by Wazbat** [comment]
Is there an agreed upon way to implement this? I'm also very interested in this feature

My use case for example is for a user to send a post request, with a json body. There would be some basic transformation using an array of data provided as a property. These requests would now all be forwarded on to the same microservice. The resulting requests would then be aggregated together into one array and sent back

**Comment by alok2k5ingh** [comment]
is aggregator feature not implemented yet?

**Comment by segara2410** [comment]
Krakend already supports this feature https://www.krakend.io/docs/backends/, even in sequential way https://www.krakend.io/docs/endpoints/sequential-proxy/
